Add DeltaFIFO (a controller framework piece)

lavalamp opened the pull request:
Splitting off completed chunks from #5270.

lavalamp committed:
Update FIFO documentation

lavalamp committed:
Add a "delta fifo", which queues up object changes

Test coverage for module at 80%.

lavalamp committed:
Add ListKeys() to FIFO & test

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+// KeyOf exposes f's keyFunc, but also detects the key of a Deltas object or
+// DeletedFinalStateUnknown objects.
+func (f *DeltaFIFO) KeyOf(obj interface{}) (string, error) {
+	if d, ok := obj.(Deltas); ok {
```
perhaps a type switch here instead of two casts + check?

lavalamp replied:
nope-- has to be done serially because the deletion marker would be stored as a delta...

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+	}
+	id, err := f.KeyOf(deltas.Newest().Object)
+	if err != nil {
+		return fmt.Errorf("couldn't create key for object: %v", err)
```
hrm, you're eating/obscuring the upstream error here.  I have a preference for either returning that error directly, or returning an error type that you can grab the error from.

fmt.Errorf is convenient, but also makes upstream debugging/checking annoying.

lavalamp replied:
I agree about the errors. It's a problem that should be fixed for the current FIFO at the same time.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) error {
+	id, err := f.KeyOf(obj)
+	if err != nil {
+		return fmt.Errorf("couldn't create key for object: %v", err)
```
as above wrt to masking the error.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+		newDeltas = f.deltaCompressor.Compress(newDeltas)
+	}
+	if len(newDeltas) > 0 {
+		if _, exists := f.items[id]; !exists {
```
given that this is a near-clone of 127-134 above, refactor into a helper function?

lavalamp replied:
I think it's only superficially similar, unfortunately.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+		}
+		f.items[id] = newDeltas
+		f.cond.Broadcast()
+	} else if _, exists := f.items[id]; exists {
```
this test is duped in both places in this condition, if you don't do the refactor suggested above, perhaps extract the check to above the if?

lavalamp replied:
Not sure it's an improvement, but ok.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+}
+
+// List returns a list of all the items; it returns the object
+// from the most recent Delta.
```
Comment on thread safety?  ("The slice is clean, but the contained interfaces are references")

brendandburns commented:
Part way done, will look more later.

yujuhong commented:
Re-assigning to Brendan since he has already started reviewing (it was gonna take me a while to catch up anyway).

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+func (f *DeltaFIFO) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	key, err := f.KeyOf(obj)
+	if err != nil {
+		return nil, false, fmt.Errorf("couldn't create key for object: %v", err)
```
same, wrt masking errors

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+	return f.GetByKey(key)
+}
+
+// GetByKey returns the complete list of deltas for the requested item,
```
I'm not sure that this comment is correct, since the empty list is the correct list for a missing item, and you are setting exists no matter what.

lavalamp replied:
'or' is inclusive, so this is technically correct... the best kind of correct. However, I'll change it.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+	return item, exists, nil
+}
+
+// Pop waits until an item is ready and returns it. If multiple items are
```
Perhaps hilight a little more that this blocks if the queue is empty.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+	for _, item := range list {
+		key, err := f.KeyOf(item)
+		if err != nil {
+			return fmt.Errorf("couldn't create key for object: %#v", err)
```
as above wrt wrapping errors.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+// TODO: consider merging keyLister with this object, tracking a list of
+//       "known" keys when Pop() is called. Have to think about how that
+//       affects error retrying.
+func NewDeltaFIFO(keyFunc KeyFunc, compressor DeltaCompressor, knownObjectKeys KeyLister) *DeltaFIFO {
```
Move this to the top of the file, since it's arguably the most "external" method?

bgrant0607 replied:
+1

brendandburns replied:
Can you address this comment?

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+}
+
+// DeltaCompressor is an algorithm that removes redundant changes.
+type DeltaCompressor interface {
```
now that I think about it, I'm not sure what this interface buys you over the func typedef.  I'd delete this interface.

lavalamp replied:
It lets you skip a helper function if you have an object that already has the right method; also it's self-documenting that your object Foo implements Xer with the DoX() method. It's the pattern the standard library uses.

brendandburns replied:
But with the function typedef you can just pass foo.Function instead of passing the full foo, I think that is actually much clearer.  There's no interface that your object implements.  It just has a function that implements that interface.

lavalamp replied:
I maintain that this code is consistent with http://golang.org/pkg/net/http/#Handler, and I provide the function adapter below, so it's easy enough to use however is necessary. On a scale of 1 - 10, my preference for leaving this is about a 7.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+// nil if there are no deltas.
+func (d Deltas) Oldest() *Delta {
+	if len(d) > 0 {
+		return &d[0]
```
This is sketchy in the context of slice growth/change.  We should at least annotate the thread semantics, and consider making the slice be of ```*Deltas``` instead of ```Deltas```

lavalamp replied:
Pop() removes from the queue, so there's only ever one owner. Arguably Get() is problematic.

brendandburns commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+// nil if there are no deltas.
+func (d Deltas) Newest() *Delta {
+	if n := len(d); n > 0 {
+		return &d[n-1]
```
here too.

brendandburns commented:
Nothing too major.

lavalamp commented:
Will make changes and do another push when I get back next week, couldn't resist arguing a few points though. ;)

bgrant0607 commented on pkg/client/cache/delta_fifo.go:
```diff
@@ -0,0 +1,391 @@
+
+// DeltaFIFO is like FIFO, but allows you to process deletes.
+//
+// DeltaFIFO solves this use case:
```
More documentation of the model and intended usage is necessary.

bgrant0607 replied:
This is basically a producer-consumer workqueue implementation?

bgrant0607 commented on pkg/client/cache/delta_fifo_test.go:
```diff
@@ -0,0 +1,277 @@
+		testFifoObjectKeyFunc,
+		// This function just keeps the most recent delta
+		// and puts deleted ones in the list.
+		DeltaCompressorFunc(func(d Deltas) Deltas {
```
What do you imagine the "compressor" doing?

lavalamp replied:
I imagine it taking a series of Updates and dropping all but the most recent. It's up to the user because I can think of different ways to combine things depending on what I want to do.

lavalamp committed:
Add proper key error type

lavalamp commented:
Mostly fixed. Still thinking about Get/List & thread safety.

lavalamp commented:
OK, I pushed a change to make sure that all slices (the Deltas objects) always only have one owner, which should resolve the potential threading problem that @brendandburns noticed. Sorry for the delay, should be ready to go now.

brendandburns commented:
Aside from moving the function up like @bgrant0607 and I requested, LGTM.  Defer to Brian for final LGTM that his comments have been addressed.  Move the function up and get an LGTM from @bgrant0607 and I'll merge.

--brendan

lavalamp commented:
Huh, I thought I did move that--

Oh, I see, I moved it *almost* all the way to the top, I thought it was useful to see the type definition first still. I don't care that much, I moved it the rest of the way. PTAL

lavalamp commented:
Travis is a flake, filed #6233
